contributing/topics/schema-design-considerations.md

```diff
@@ -0,0 +1,152 @@
+# Schema Design Considerations
+
+Whilst it is acceptable in certain cases to map the schema of a new resource or feature when extending an existing resource, one-to-one from the Azure API, in the majority of cases more consideration needs to be given how to expose the Azure API in Terraform so that the provider presents a consistent and intuitive experience to the end user.
+
+Below are a list of common patterns found in the Azure API and how these typically get mapped within Terraform.
+
+## Features that are toggled by the property `Enabled`
+
+It is commonplace for features to be toggled on and off by an `Enabled` property within an object in the SDK used to interact with the Azure API. See the examples below.
+
+Example A.
+```go
+type ManagedClusterStorageProfileBlobCSIDriver struct {
+	Enabled *bool `json:"enabled,omitempty"`
+}
+```
+
+Example B.
+```go
+type ManagedClusterWorkloadAutoScalerProfileVerticalPodAutoscaler struct {
+	ControlledValues ControlledValues `json:"controlledValues"`
+	Enabled          bool             `json:"enabled"`
+	UpdateMode       UpdateMode       `json:"updateMode"`
+}
+```
+
+Although there are still resources within the provider where this property is exposed in the Terraform schema, the provider is moving away from this and instead translates this behaviour in one of two ways.
+
+In cases where `Enabled` is the only field required to turn a feature on and off we opt to flatten the block into a single top level property (or higher level property if already nested inside a block). So in the case of Example A, this would become:
+
+```go
+"storage_blob_driver_enabled": {
+    Type:     pluginsdk.TypeBool,
+    Optional: true,
+    Default:  false,
+},
+```
+
+For features that can accept or require configuration, i.e. the object contains additional properties other than `Enabled` like in Example B, the behaviour should be such that when the block is present the feature is enabled, and when it is absent it is disabled. The corresponding Terraform schema would be as follows:
+
+```go
+"vertical_pod_autoscaler": {
+    Type:     pluginsdk.TypeList,
+    Optional: true,
+    MaxItems: 1,
+    Elem: &pluginsdk.Resource{
+        Schema: map[string]*pluginsdk.Schema{
+            "update_mode": {
+                Type:     pluginsdk.TypeString,
+                Required: true,
+                ValidateFunc: validation.StringInSlice([]string{
+                    string(managedclusters.UpdateModeAuto),
+                    string(managedclusters.UpdateModeInitial),
+                    string(managedclusters.UpdateModeOff),
```

Comment:
We shouldn't be exposing `Off` here, this probably wants to become Optional (with no default value) - then specify we default this to `Off` in the expand function?

Reply:
I see your point, I chose to remove the `Off` value entirely for this "hypothetical" example since I think it would overcomplicate and cloud the point that is trying to be made by this example. Instead I chose to expand on it under the `None` case since I think that is more relevant. WDYT?